File: Android/app/src/main/java/com/whatsapp/otp/sample/app/receiver/OtpCodeReceiver.java
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.whatsapp.otp.sample.app.receiver;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import com.whatsapp.otp.android.sdk.WhatsAppOtpIncomingIntentHandler;
import com.whatsapp.otp.client.WaOtpUtils;
import com.whatsapp.otp.common.WaLogger;
import com.whatsapp.otp.sample.app.fragment.OtpValidatorFragment;
import com.whatsapp.otp.sdkextension.WhatsAppHandshakeHandler;
import dagger.hilt.android.AndroidEntryPoint;
import javax.inject.Inject;

@AndroidEntryPoint
public class OtpCodeReceiver extends BroadcastReceiver {

  private static final WaLogger WA_LOGGER = WaLogger.getLogger(OtpCodeReceiver.class);
  private static final String REQUEST_ID_KEY = "request_id";

  @Inject
  WhatsAppOtpIncomingIntentHandler incomingIntentHandler;

  @Inject
  WhatsAppHandshakeHandler handshakeHandler;

  @Override
  public void onReceive(final Context context, final Intent intent) {
    WA_LOGGER.info("Code received");
    final String requestId = intent.getStringExtra(REQUEST_ID_KEY);
    WA_LOGGER.info("Handshake Request id: " + requestId);

    final String expectedHandshakeId = getExpectedHandshakeId(requestId);
    if (expectedHandshakeId == null) {
      WA_LOGGER.error("Failed to login. Reason: Handshake ID not expected");
      return;
    }

    incomingIntentHandler.processOtpCode(intent,
        expectedHandshakeId,
        (code) -> broadcastCodeToInternalComponents(context, code),
        (whatsAppOtpError, e) -> WA_LOGGER.error(
            "Failed to login. Reason:" + whatsAppOtpError.name()));
  }

  private String getExpectedHandshakeId(final String receivedHandshakeId) {
    if (receivedHandshakeId == null) {
      return null;
    }
    return handshakeHandler.getExpectedHandshakeId(receivedHandshakeId);
  }

  private static void broadcastCodeToInternalComponents(final Context context, final String code) {
    Intent loginIntent = WaOtpUtils.createInternalCodeBroadcasterIntent(code,
        OtpValidatorFragment.OTP_CODE_RECEIVER,
        context);
    context.sendBroadcast(loginIntent);
  }
}

File: WhatsApp OTP Sample Mobile App/ReactNative/android/app/src/main/java/com/whatsapp/otp/reactnative/OtpCodeReceiver.kt
package com.whatsapp.otp.reactnative

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent

class OtpCodeReceiver : BroadcastReceiver() {

    companion object {
        private const val OTP_CODE_KEY = "code"
        private const val OTP_ERROR_KEY = "error"
        const val OTP_RETRIEVED_ACTION = "com.whatsapp.otp.OTP_RETRIEVED"
    }

    override fun onReceive(context: Context?, intent: Intent?) {
        if (intent?.action == OTP_RETRIEVED_ACTION) {
            val code = intent.getStringExtra(OTP_CODE_KEY)
            val error = intent.getStringExtra(OTP_ERROR_KEY)

            val internalIntent = Intent(WhatsAppCodeReceiverActivity.OTP_RETRIEVED_ACTION).apply {
                code?.let { putExtra(OTP_CODE_KEY, it) }
                error?.let { putExtra(OTP_ERROR_KEY, it) }
            }

            context?.sendBroadcast(internalIntent)
        }
    }
}
